### PR TITLE
507-Refacto Code for NetworkDialog And add check on Url

### PR DESCRIPTION
**Pull request by Esquimor**
It is currently in: Work In Progress.

Link issue: https://github.com/archethic-foundation/archethic-wallet/issues/507

Ask for this issue:

- Check if url was correctly formated (For me, it was already the case)
- Check if url enter by the user was in the list of used url by the network.

What I have done:

- Refacto code to split in multiples components
- Improve readability of "onPressed" to check custom url

What blocks me:

- When I try to get the nodes. I have an error: ArchethicConnectionException
`responseHttp.error={
"linkException": "ServerException(originalException: Connection refused, originalStackTrace: #0      IOClient.send (package:http/src/io_client.dart:88:7)\n<asynchronous suspension>\n#1      HttpLink._executeRequest (package:gql_http_link/src/link.dart:131:24)\n<asynchronous suspension>\n#2      HttpLink.request (package:gql_http_link/src/link.dart:76:26)\n<asynchronous suspension>\n#3      Stream.first.<anonymous closure> (dart:async/stream.dart:1586:25)\n<asynchronous suspension>\n, parsedResponse: null)"`

### CODE REVIEW

**Comment by Esquimor** [comment]
I have found how take nodes from network.

I need to custom errors messages and clean the code.

**Comment by Esquimor** [comment]
@redDwarf03 you can review

File: lib/ui/widgets/dialogs/network_dialog.dart [hunk 1]
```diff
@@ -76,145 +64,110 @@ class NetworkDialog {
...
-                                                  networkDevEndpoint:
-                                                      endpointController.text,
-                                                ),
-                                              );
-
-                                          Navigator.pop(context);
-                                        }
-                                      }
-                                    },
+                        return _NetworkDialogCustomInput(
```
Comment (Chralu) [inline lib/ui/widgets/dialogs/network_dialog.dart:82]: Great refacto 👍

File: lib/ui/widgets/dialogs/network_dialog.dart [hunk 2]
```diff
@@ -76,145 +64,110 @@ class NetworkDialog {
...
+                            void setError(String errorText) {
+                              setState(() {
+                                endpointError = errorText;
+                                FocusScope.of(context).requestFocus(
+                                  endpointFocusNode,
+                                );
+                              });
+                            }
+
+                            Future<List<Node>> getNodeFromNetwork(
```
Comment (Chralu) [inline lib/ui/widgets/dialogs/network_dialog.dart]: These business functions could be Family Async Providers declared in a new `/lib/application/network` directory.


Example using `Riverpod` generator.

<details>
<summary>
/lib/application/network/provider.dart
</summary>

```dart
@riverpod
List<Node> _networkNodes(
  _NetworkNodesRef ref, {
  required AvailableNetworks network,
}) {
    final networkSetting = NetworksSetting(
      network: network,
      networkDevEndpoint: '',
    );

    final link = networkSetting.getLink();

    final nodeListMain = await ApiService(
      link,
    ).getNodeList();

    return nodeListMain;
}

@riverpod
bool _isReservedNodeUri(
  _IsReservedNodeUri ref, {
  required Uri uri,
}) {
  final nodeListMain = await ref.watch(_networkNodesProvider(network AvailableNetworks.archethicMainNet));
  final nodeListTest = await ref.watch(network: _networkNodesProvider(AvailableNetworks.archethicTestNet));

  return nodeListMain.followedBy(nodeListTest).any(
      (node) =>
          node.ip == uri.host &&
          node.port == uri.port,
    );
}

abstract class NetworkProvider {
  static final isReservedNodeUri = _isReservedNodeUriProvider;
}
```
</details>


<details>
<summary>
NetworkDialog
</summary>

```dart
....
   onSubmitNetwork: () async {
      if (await ref.readNetworkProvider.isReservedNodeUri(uriInput)) {
          setError(
            localizations.enterEndpointUseByNetwork,
          );
          return;
       }

        await ref
            .read(
              SettingsProviders.settings.notifier,
            )
            .setNetwork(
              NetworksSetting(
                network: AvailableNetworks.archethicDevNet,
                networkDevEndpoint: textCleaned,
              ),
            );
   }
...
```
</details>

**Review by Chralu — APPROVED** [review]

**Comment by redDwarf03** [comment]
Hello
My feedbacks
- increase the length of the field because it's too small for an url (200 characters)
- i prefer to have a snackbar for error message please
```
     UIUtil.showSnackbar(
                              endpointError,
                              context,
                              ref,
                              theme.text!,
                              theme.snackBarShadow!,
                            );
```
- it's possible to fill "https://mainnet.archethic.net" for example. With IP Node, could you exclude domain name "testnet.archethic.net" and "mainnet.archethic.net" please
- When user fills an domain name or IP and validates, when he come back on network dialog, the old value is forgot and the form is initialized with http://localhost:4000

**Comment by Esquimor** [comment]
@redDwarf03 

I agree to add "testnet.archethic.net" and "mainnet.archethic.net" in the value to check and increase the size of the input.

But, the others points, I will probably add new issue, related to this one:

- Probably, the error message need to be check in all the application.
- And, keep the value enter, by the user, is out of scope for this issue.

**Comment by redDwarf03** [comment]
> @redDwarf03
> 
> I agree to add "testnet.archethic.net" and "mainnet.archethic.net" in the value to check and increase the size of the input.
> 
> But, the others points, I will probably add new issue, related to this one:
> 
> * Probably, the error message need to be check in all the application. -> please, implement just for this case the snackbar but after that, i will create an issue to manage error message more globally in the app
> * And, keep the value enter, by the user, is out of scope for this issue. -> ok

**Comment by redDwarf03** [comment]
Thank you for your contribution and your work. I merge it

**Pull request merged**